src/meepgeom.cpp

```diff
@@ -1070,33 +1133,107 @@ void geom_epsilon::eff_chi1inv_matrix(meep::component c, symm_matrix *chi1inv_ma
     return;
   }
 
-  if (!get_front_object(v, geometry_tree, p, &o, shiftby, mat, mat_behind)) {
-    get_material_pt(mat, v.center());
-    if (mat && (mat->which_subclass == material_data::MATERIAL_USER ||
-                mat->which_subclass == material_data::MATERIAL_GRID)
-        && mat->do_averaging) {
+  // For variable materials with do_averaging == true, switch over to slow fallback integration method.
+  // For material grids, however, we have to do some more logic first...
+  if ((is_variable(mat) && mat->do_averaging) || (is_variable(mat_behind) && mat_behind->do_averaging)) {
+    if ((!is_material_grid(mat)) && (!is_material_grid(mat_behind))){
       fallback = true;
       return;
     }
-    else {
+  }
+
+  /* if we have a material grid, we need to calculate the fill fraction,
+  normal vector, etc. we also need to determine if we really need
+  to do any averaging, or if the current pixel is smooth enough
+  to just evaluate.
+  
+  Keep in mind that if both the front and back materials are material grids,
+  we assume they are the same material grid. (all of the complicated overlapping
+  symmetry conditions are taken care of elsewhere).
+   */
+  double fill_front, fill_back;
+  vector3 normal_front, normal_back;
+  if (is_material_grid(mat)){
+    int oi;
+    geom_box_tree tp = geom_tree_search(p_mat, restricted_tree, &oi);
+    
+    meep::vec normal_vec_front(matgrid_grad(p_mat, tp, oi, mat));
+    
+    double nabsinv = 1.0 / meep::abs(normal_vec_front);
+    LOOP_OVER_DIRECTIONS(normal_vec_front.dim, k) { normal_vec_front.set_direction(k,normal_vec_front.in_direction(k)*nabsinv); }
+    
+    double uval = matgrid_val(p_mat, tp, oi, mat)+this->u_p;
+    double d = (mat->eta-uval) * nabsinv;
+    double r = v.diameter()/2;
+    
+    fill = get_material_grid_fill(normal_vec_front.dim,d,r,uval,mat->eta,&mg_averaging,&center_is_void);
+    normal = vec_to_vector3(normal_vec_front);
+  }
+  if (is_material_grid(mat_behind)){
```

Comment:
Should be no need for this.  If you have two overlapping objects, both with a material grid, then you should treat the material grids as if they were each *uniform* materials and fall back on our geometric-object fill fraction, similar to variable materials.